gvfs-helper.c

```diff
@@ -0,0 +1,2074 @@
+
+	if (!file_exists(buf_path.buf) && mkdir(buf_path.buf, 0777) == -1) {
+		error("cannot create directory for loose object '%s'",
+		      buf_path.buf);
+		goto cleanup;
+	}
+
+	strbuf_addch(&buf_path, '/');
+	strbuf_addstr(&buf_path, hex+2);
+	strbuf_addstr(&buf_path, ".temp");
+
+	tempfile = create_tempfile(buf_path.buf);
+	if (!tempfile)
+		goto cleanup;
+
+	fdopen_tempfile(tempfile, "w");
+
+cleanup:
+	strbuf_release(&buf_path);
+
+	return tempfile;
+}
+
+/*
+ * Convert the tempfile into a permanent .pack packfile in the ODB.
+ * Create the corresponding .idx file.
+ *
+ * Return the full pathname of the resulting packfile.
+ */
+static enum gh__error_code install_packfile(struct tempfile **pp_tempfile,
+					    struct strbuf *packfile_name)
+{
+	struct child_process ip = CHILD_PROCESS_INIT;
+	struct strbuf pack_name_tmp = STRBUF_INIT;
+	struct strbuf pack_name_dst = STRBUF_INIT;
+	struct strbuf idx_name_tmp = STRBUF_INIT;
+	struct strbuf idx_name_dst = STRBUF_INIT;
+	size_t len_base;
+	enum gh__error_code ec = GH__ERROR_CODE__OK;
+
+	strbuf_setlen(packfile_name, 0);
+
+	/*
+	 * start with "<base>.temp" (that is owned by tempfile class).
+	 * rename to "<base>.pack.temp" to break ownership.
+	 *
+	 * create "<base>.idx.temp" on provisional packfile.
+	 *
+	 * officially install both "<base>.{pack,idx}.temp" as
+	 * "<base>.{pack,idx}".
+	 */
+
+	strbuf_addstr(&pack_name_tmp, get_tempfile_path(*pp_tempfile));
+	if (!strip_suffix(pack_name_tmp.buf, ".temp", &len_base)) {
+		/*
+		 * This is more of a BUG(), but I want the error
+		 * code propagated.
+		 */
+		error("packfile tempfile does not end in '.temp': '%s'",
+		      pack_name_tmp.buf);
+		ec = GH__ERROR_CODE__COULD_NOT_INSTALL_PACKFILE;
+		goto cleanup;
+	}
+
+	strbuf_setlen(&pack_name_tmp, (int)len_base);
+	strbuf_addbuf(&pack_name_dst, &pack_name_tmp);
+	strbuf_addbuf(&idx_name_tmp, &pack_name_tmp);
+	strbuf_addbuf(&idx_name_dst, &pack_name_tmp);
+
+	strbuf_addstr(&pack_name_tmp, ".pack.temp");
+	strbuf_addstr(&pack_name_dst, ".pack");
+	strbuf_addstr(&idx_name_tmp, ".idx.temp");
+	strbuf_addstr(&idx_name_dst, ".idx");
+
+	// TODO if either pack_name_dst or idx_name_dst already
+	// TODO exists in the ODB, create alternate names so that
+	// TODO we don't step on them.
+
+	if (rename_tempfile(pp_tempfile, pack_name_tmp.buf) == -1) {
+		error("could not rename packfile to '%s'", pack_name_tmp.buf);
+		ec = GH__ERROR_CODE__COULD_NOT_INSTALL_PACKFILE;
+		goto cleanup;
+	}
+
+	argv_array_push(&ip.args, "index-pack");
+	if (gh__cmd_opts.show_progress)
+		argv_array_push(&ip.args, "-v");
+	argv_array_pushl(&ip.args, "-o", idx_name_tmp.buf, NULL);
+	argv_array_push(&ip.args, pack_name_tmp.buf);
+	ip.git_cmd = 1;
+	ip.no_stdin = 1;
+	ip.no_stdout = 1;
+
+	// TODO consider capturing stdout from index-pack because
+	// TODO it will contain the SHA of the packfile and we can
+	// TODO (should?) add it to the .pack and .idx pathnames
+	// TODO when we install them.
```

Comment:
It would be best to add it to the names, if possible.

Reply:
Keep in mind that it's the last 20 bytes of the .pack, but the last 21-40 bytes of the .idx file.